### PR TITLE
Add server side rendering and static pre-rendering of barista.

### PR DESCRIPTION
**Pull request by lukasholzer**
### <strong>Pull Request</strong>

<hr>
Hi, thank you for contributing to Barista with this pull request (PR).

To ensure a fast process and merging of your PR please make sure it fulfills the
coding standards and contribution guidelines.

- A feature proposal has been provided, discussed and approved first.
- There is a meaningful description of the issue in GitHub (Screenshots are
  often helpful).
- If the PR introduces breaking-changes or deprecations it matches the following
  guidelines.
  - The commit message follows our commit guidelines.
  - Tests for the changes have been added (for bug fixes / features).
  - Docs have been added / updated (for bug fixes / features).

Please choose the type appropriate for the changes below: <br>

#### Checklist

- [ ] I have read the CONTRIBUTING doc and I follow the PR guidelines
- [ ] Lint and unit tests pass locally with my changes
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)

### CODE REVIEW

File: environments/barista-environment.ts [hunk 1]
```diff
@@ -14,17 +14,22 @@
...
  */
 
-import { join } from 'path';
 import { isPublicBuild } from '@dynatrace/tools/shared';
+import { config as dotenvConfig } from 'dotenv';
+import { join } from 'path';
 import { BaEnvironment } from './barista-environment.interface';
 
+// TODO: Remove all platform dependent usages like process and dotenv
+// load the environment variables from the .env file in your workspace
```
Comment (ffriedl89) [inline environments/barista-environment.ts]: I am confused :) should we use dotenv and a .env file now or not?

Comment (lukasholzer) [reply environments/barista-environment.ts]: this is not a file consumed by angular :) only from nodejs. but i don't like it to access the process.env but we do it already so I added a todo to remove it later on I prefer the approach to add a barista-environment.internal.ts that is generated by our jenkins. But this should be part of a different issue

File: apps/barista-design-system/src/shared/services/toc.service.ts [hunk 2]
```diff
@@ -76,27 +77,31 @@ export class BaTocService implements OnDestroy {
...
-          for (const tocItem of this.tocItems) {
-            const scrollItemId = scrollItem.element.getAttribute('id');
-            if (tocItem.headlineId === scrollItemId) {
-              this.activeItems.next([tocItem]);
-            }
-            for (const tocSubItem of tocItem.subheadlines) {
-              if (tocSubItem.headlineId === scrollItemId) {
-                this.activeItems.next([tocItem, tocSubItem]);
+    // TODO: whole scroll spy logic has to be fixed! It was not refactored
+    // during the angular router refactoring
```
Comment (tomheller) [inline apps/barista-design-system/src/shared/services/toc.service.ts]: Can you please create a follow up issue for that one, @lukasholzer ?

File: apps/barista-design-system/src/shared/services/universal.interceptor.ts [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+} from '@angular/common/http';
+import { Inject, Injectable, Optional } from '@angular/core';
+import { REQUEST } from '@nguniversal/express-engine/tokens';
+import { Request } from 'express';
+import { Observable } from 'rxjs';
+
+/**
+ * In a Universal app, HTTP URLs must be absolute.
+ * This means we have to use absolute URLs when running on the server
+ * and relative URLs when running in the browser.
```
Comment (tomheller) [inline apps/barista-design-system/src/shared/services/universal.interceptor.ts]: Why do they have to be relative in the browser? They should work as absolute paths in the browser as well, right?

**Comment by lukasholzer** [comment]
image

File: environments/barista-environment.ts (hunk 1 above)
Comment (tomheller) [reply environments/barista-environment.ts]: I agree with @lukasholzer here, we should try and move everything that was configured in the process.env to the .env file. 
We just need to make sure that the .env file is not within the git index.

File: libs/workspace/src/builders/barista-build/schema.json [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+    },
+    "routesFile": {
+      "type": "string",
+      "description": "Path to the file that holds the route information."
+    },
+    "routes": {
+      "type": "string",
+      "description": "Comma separated list of routes."
+    }
+  }
```
Comment (tomheller) [inline libs/workspace/src/builders/barista-build/schema.json]: `outputPath` and `devServerTartet` are not marked optional in your interface, you should mark them as requiredAttributes here. 
routesFile and routes are marked as optional but do not define a default value here. Maybe add a description that either the routesFile or the routes are necessary

File: apps/barista-design-system/src/shared/services/universal.interceptor.ts (hunk 3 above)
Comment (lukasholzer) [reply apps/barista-design-system/src/shared/services/universal.interceptor.ts]: universal rendering needs this http interceptor:

https://angular.io/guide/universal#using-absolute-urls-for-server-requests

File: libs/workspace/src/schematics/workspace/files/src/index.ts.template [hunk 5]
```diff
@@ -0,0 +1 @@
+const variable = "<%= projectName %>";
```
Comment (ffriedl89) [inline libs/workspace/src/schematics/workspace/files/src/index.ts.template]: dummy?

File: libs/workspace/src/schematics/workspace/schema.d.ts [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+export interface WorkspaceSchematicSchema {
+  name: string;
+  tags?: string;
+  directory?: string;
+}
```
Comment (ffriedl89) [inline libs/workspace/src/schematics/workspace/schema.d.ts]: i guess this is also generated by the schematic and can be removed

File: apps/barista-design-system/src/shared/services/toc.service.ts (hunk 2 above)
Comment (lukasholzer) [reply apps/barista-design-system/src/shared/services/toc.service.ts]: done: #796

File: libs/workspace/src/schematics/workspace/schema.json [hunk 7]
```diff
@@ -0,0 +1,28 @@
+{
```
Comment (ffriedl89) [inline libs/workspace/src/schematics/workspace/schema.json]: same for this file?

File: libs/workspace/src/schematics/workspace/schematic.spec.ts [hunk 8]
```diff
@@ -0,0 +1,42 @@
+/**
```
Comment (ffriedl89) [inline libs/workspace/src/schematics/workspace/schematic.spec.ts]: remove

File: libs/workspace/src/schematics/workspace/schematic.ts [hunk 9]
```diff
@@ -0,0 +1,103 @@
+/**
```
Comment (ffriedl89) [inline libs/workspace/src/schematics/workspace/schematic.ts]: remove?

File: package.json [hunk 10]
```diff
@@ -89,10 +90,11 @@
     "@nguniversal/builders": "^9.0.0",
     "@nguniversal/express-engine": "^9.0.0",
     "@nguniversal/module-map-ngfactory-loader": "^9.0.0-next.9",
-    "@nrwl/angular": "^9.1.2",
-    "@nrwl/jest": "^9.1.2",
-    "@nrwl/node": "^9.1.2",
-    "@nrwl/workspace": "^9.1.2",
+    "@nrwl/angular": "^9.0.4",
+    "@nrwl/jest": "^9.0.4",
+    "@nrwl/node": "^9.0.4",
```
Comment (ffriedl89) [inline package.json]: can be 9.1.2 ?

File: package.json [hunk 11]
```diff
@@ -89,10 +90,11 @@
...
+    "@nrwl/nx-plugin": "^9.1.0",
+    "@nrwl/workspace": "^9.0.4",
```
Comment (ffriedl89) [inline package.json]: can be 9.1.2?

File: libs/workspace/src/builders/barista-build/schema.json (hunk 4 above)
Comment (lukasholzer) [reply libs/workspace/src/builders/barista-build/schema.json]: the required was not working so I removed it. `routesFile` and routes are optional and have a fallback in the code. You cannot provide default values for it.

image

image

**Pull request merged**